Z_Algorithm/README.md

```diff
@@ -0,0 +1,71 @@
+# Z algorithm
+This algorithm finds all occurrences of a pattern in a text in linear time. Let length of text be n and of pattern be m, then total time taken is O(m + n) with linear space complexity. Now we can see that both time and space complexity is same as KMP algorithm but this algorithm is Simpler to understand.
+Z algorithm is a linear time string matching algorithm which runs in  complexity. It is used to find all occurrence of a pattern  in a string , which is common string searching problem.
+In this algorithm, we construct a Z array.
+
+### What is Z Array?
+For a string str[0..n-1], Z array is of same length as string. An element Z[i] of Z array stores length of the longest substring starting from str[i] which is also a prefix of str[0..n-1]. The first entry of Z array is meaning less as complete string is always prefix of itself.
+
+```
+Example:
+
+Index            0   1   2   3   4   5   6   7   8   9  10  11 
+Text             a   a   b   c   a   a   b   x   a   a   a   z
+Z values         X   1   0   0   3   1   0   0   2   2   1   0
+
+str  = "aaaaaa"
+Z[]  = {x, 5, 4, 3, 2, 1}
+
+str = "aabaacd"
+Z[] = {x, 1, 0, 2, 1, 0, 0}
+
+str = "abababab"
+Z[] = {x, 0, 6, 0, 4, 0, 2, 0}
+```
+
+### How is Z array helpful in Searching Pattern in Linear time?
+The idea is to concatenate pattern and text, and create a string “P$T” where P is pattern, $ is a special character should not be present in pattern and text, and T is text. Build the Z array for concatenated string. In Z array, if Z value at any point is equal to pattern length, then pattern is present at that point.
+```
+Example:
+Pattern P = "aab",  Text T = "baabaa"
+The concatenated string is = "aab$baabaa"
+Z array for above concatenated string is {x, 1, 0, 0, 0, 
+                                          3, 1, 0, 2, 1}.
+Since length of pattern is 3, the value 3 in Z array 
+indicates presence of pattern. 
+```
+
+### How to construct Z array?
+A Simple Solution is two run two nested loops, the outer loop goes to every index and the inner loop finds length of the longest prefix that matches substring starting at current index. 
+
+# ALGORITHM
```

Comment:
This a huge description . remove it and provide some simple steps in bullet form